src/persist-txn/src/txn_read.rs

```diff
@@ -53,11 +53,55 @@ pub struct DataSnapshot<T> {
     pub(crate) latest_write: Option<T>,
     /// The as_of asked for.
     pub(crate) as_of: T,
-    /// An upper bound on the times known to be empty of unapplied writes via txns.
-    pub(crate) empty_to: T,
 }
 
 impl<T: Timestamp + Lattice + TotalOrder + Codec64> DataSnapshot<T> {
+    /// See [SinceHandle::snapshot_stats].
+    pub fn snapshot_stats<K, V, D, O>(
+        &self,
+        data_since: &SinceHandle<K, V, T, D, O>,
+    ) -> impl Future<Output = Result<SnapshotStats, Since<T>>> + Send + 'static
+    where
+        K: Debug + Codec + Ord,
+        V: Debug + Codec + Ord,
+        D: Semigroup + Codec64 + Send + Sync,
+        O: Opaque + Codec64,
+    {
+        // This is used by the optimizer in planning to get cost statistics, so
+        // it can't use `unblock_read`. Instead use the "translated as_of"
+        // trick we invented but didn't end up using for read of the shard
+        // contents. The reason we didn't use it for that was because we'd have
+        // to deal with advancing timestamps of the updates we read, but the
+        // stats we return here don't have that issue.
+        //
+        // TODO: If we don't have a `latest_write`, then the `None` option to
+        // `snapshot_stats` is not quite correct because of pubsub races
+        // (probably marginal) and historical `as_of`s (probably less marginal
+        // but not common in mz right now). Fixing this more precisely in a
+        // performant way (i.e. no crdb queries involved) seems to require that
+        // persist-txn always keep track of the latest write, even when it's
+        // known to have been applied. `snapshot_stats` is an estimate anyway,
+        // it doesn't even attempt to account for things like consolidation, so
+        // this seems fine for now.
+        let as_of = self.latest_write.clone().map(Antichain::from_elem);
+        data_since.snapshot_stats(as_of)
+    }
```

Comment:
This is unchanged but moved to a different `impl` block without the `StepForward` trait bound.